### PR TITLE
Support cut, copy and paste on open shadow DOM inputs

### LINKED ISSUE #1026: Support open shadow DOM components
### Problem description

Hello,

this is a follow up to #1025. Currently the library doesn't support web components which use shadow dom. This feature request only deals with **open** shadow doms. The underlying issue seems to be that with web components the element that is focused can be different from the one where input needs to take place: Web Components wrap builtin elements on which the input takes place. To get the element to interact with `document.activeElement` seems to be used. In case of web components this would return the web component, while the element that would need to be interacted with (e. g. an input element within the shadow DOM) is hidden from the library. 

 

### Suggested solution

There has been prior art in the form of `getActiveElement` (https://github.com/testing-library/user-event/blob/main/src/utils/focus/getActiveElement.ts) which iterates over potential shadow DOM's `activeElement` property, until the element to interact with is reached. 

This could be used as a replacement for `document.activeElement` to make the library open shadow DOM compatible. I quickly hacked this together on `paste()` and it seems to get the job done: 
https://github.com/Christian24/user-event/blob/f4242842c1ed51d2ec8042263bee7abdbafea21e/src/clipboard/paste.ts#L14

In some cases like keyboard input it seems to be more an issue of getting the element to focus properly first. 

Interestingly, `user.click(element)` doesn't seem to focus the input within the web component, not sure why. I had to manually `querySelector` the input to get my paste test to work: 
https://github.com/Christian24/user-event/blob/main/tests/webcomponents/index.ts#L29

### Additional context

I have been playing a little bit with this this afternoon: https://github.com/Christian24/user-event 
Not sure it will evolve into a PR though, would like to hear seasoned contributors opinion first. Also for some reason I couldn't get the tests to debug in VS Code...

### PR DESCRIPTION
**Pull request by Christian24**
**What**:
This allows support for copy, cut & paste in inputs in open shadow dom custom elements. 

**Why**:
Because I filled #1025 and #1026. This doesn't completely resolve #1026 though, however it should ease the situation when using native inputs in custom elements.   

**How**:
Changed the implementation of getting the activeElement. 

**Checklist**:

- [ ] Documentation (I am uncertain about this, maybe the state of #1026 should be at some point be part of the documentation?)
- [x ] Tests
- [ x] Ready to be merged

### CODE REVIEW

File: src/clipboard/copy.ts [hunk 1]
```diff
@@ -1,10 +1,10 @@
 import {copySelection} from '../document'
-import type {Instance} from '../setup'
-import {writeDataTransferToClipboard} from '../utils'
+import {Instance} from '../setup'
+import {getActiveElement, writeDataTransferToClipboard} from '../utils'
 
 export async function copy(this: Instance) {
   const doc = this.config.document
-  const target = doc.activeElement ?? /* istanbul ignore next */ doc.body
+  const target = getActiveElement(doc) ?? doc.activeElement ?? doc.body
```
Comment (ph-fritsche) [inline src/clipboard/copy.ts:7]:
```suggestion
const target getActiveElementOrBody(doc)
```
We should use the util that already provides this fallback.

File: tests/dom/customElement.ts [hunk 2]
```diff
@@ -84,3 +85,94 @@ test('types text inside custom element', async () => {
...
+  const inputElement = element.shadowRoot?.querySelector('input')
+  if (inputElement) {
+    await user.click(inputElement)
+  }
+  await user.keyboard('test')
+
+  expect((element as ShadowInput).value).toEqual('test')
+})
+
+test('Render open shadow DOM element - copy', async () => {
```
Comment (ph-fritsche) [inline tests/dom/customElement.ts:124]:
```suggestion
describe('on shadow DOM', () => {
  test('copy', async () => {
```

File: tests/dom/customElement.ts [hunk 3]
```diff
@@ -84,3 +85,94 @@ test('types text inside custom element', async () => {
...
+
+  const inputElement = element.shadowRoot?.querySelector('input')
+  if (inputElement) {
+    await user.click(inputElement)
+  }
+  await user.keyboard('test')
+
+  if (inputElement) {
+    setUISelection(inputElement, {anchorOffset: 0, focusOffset: 4})
+  }
```
Comment (ph-fritsche) [inline tests/dom/customElement.ts:137]: The tests should make clear which feature is tested and be narrowed down to this - i.e. no other user-event API being used. Otherwise it might be hard for the next developer to determine what is tested, what lacks tests and which tests are redundant. See the comment in this particular file above.

File: tests/dom/customElement.ts [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import userEvent from '#src'
-import {addListeners} from '#testHelpers'
+import {addListeners, setup} from '#testHelpers'
+import {setUISelection} from '#src/document'
 
 // It is unclear which part of our implementation is targeted with this test.
 // Can this be removed? Is it sufficient?
```
Comment (ph-fritsche) [inline tests/dom/customElement.ts:6]: I don't think this test suite should be extended.

**Comment by ph-fritsche** [comment]
First of all: Thanks for contributing to this library. This is much appreciated :heart: 

I think we should add tests to `tests/clipboard/*.ts` testing just the interaction that happens on the `.copy()`/`.cut()`/`.paste()`.
If we feel like `.click()` and `.keyboard()` should have additional tests to, we should add them there.
Testing the features separately makes it much easier for future developers to identify if a test breakage is expected on a change. It's really bad if you're working on something, 50 tests break and you have to spend extra time to figure out which of those 50 tests have false assumptions that you're just about to correct and which actually warn you that you're in the process of breaking something else by mistake.

Also the thing we just imply here is the internal abstractions working correctly. So this is probably where the real testing should take place to pave the way for possible other changes required by #1026.
So we need to test `util/focus/getActiveElement and the `document/copySelection` and `event/input` to properly interact with shadow DOM.
If we know about a limitation of the current change and we don't want to wait for all aspects to be resolved, this can also mean adding some tests with comments that outline what is missing.

File: src/clipboard/copy.ts (hunk 1 above)
Comment (Christian24) [reply src/clipboard/copy.ts:7]: I had tried this, but it didn't work. I will investigate a bit.

File: tests/dom/customElement.ts (hunk 4 above)
Comment (Christian24) [reply tests/dom/customElement.ts:6]: So your idea is to move these tests into copy, paste, cut test suites?

Comment (ph-fritsche) [reply tests/dom/customElement.ts:6]: The test in this file was introduced in https://github.com/testing-library/user-event/pull/320/files#diff-a1c186e56ef39b3e82d6781e7a934af052817d6ae42102d3c12368c9896e6e07

The thought there might have been to test for `.type()` what you're trying to test for `.copy()/.paste()/.cut()`, that the target for changes (the editable context) is correctly recognized to be inside the shadow tree.
Since then the implementation for `.type()` completely changed and so did the only assertion - an event screenshot.

I'd like to remove/replace this test, but therefore we need to understand which (if any) part of this test is important and which is just noise. If it's really just about recognizing the correct target, we should make sure to assert this on any API in the respective test suite. Maybe we should add a linting rule that ensures that `document.activeElement` is only used per the util that should document the edge cases - which of course we should test on it directly.

Anyway I would not add any new tests under that code comment, it just makes it harder to be confident when removing/replacing it.

Comment (Christian24) [reply tests/dom/customElement.ts:6]: I am working on moving the tests at the moment. And I am throwing in a few goodies. :)

File: src/clipboard/copy.ts (hunk 1 above)
Comment (Christian24) [reply src/clipboard/copy.ts:7]: All of these should be fixed now.

File: tests/dom/customElement.ts (hunk 3 above)
Comment (Christian24) [reply tests/dom/customElement.ts:138]: So I added support for `selection` and `focus` in `setup()`, so this is no longer necessary.

File: tests/dom/customElement.ts (hunk 4 above)
Comment (Christian24) [reply tests/dom/customElement.ts:6]: So, I moved the tests and removed the test that was there. My reasoning is this: From what I see this is not a test of open shadow dom as the input element was directly passed to `user.type()`. I adapted the simple test I had to make sure that `type()` works directly on open shadow dom elements, although testing is going to be limited since `delegatesFocus` is not implemented in jsdom. I also think we should rather test recognizing the correct target, because the rest should be identical in open shadow dom.

**Comment by Christian24** [comment]
I added some very simple tests for `getActiveElement`. So far I have only dealt with input related stuff. This is the area most relevant to me.

File: tests/utility/type.ts [hunk 5]
```diff
@@ -90,3 +92,14 @@ test('do nothing on disabled element', async () => {
 
   expect(getEvents()).toHaveLength(0)
 })
+
+describe('on shadow DOM', () => {
+  test('type into an input element', async () => {
+    const {element, user} = setup('<shadow-input></shadow-input>')
+
+    // Skip click because delegatesFocus is not implemented in jsdom
+    await user.type(element, 'test', {skipClick: true})
```
Comment (ph-fritsche) [inline tests/utility/type.ts:101]: This only works in our specific `setup` that includes the focus workaround.
We should make sure this works inside our own tests and for our users.

File: tests/utils/focus/getActiveElement.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import '../../_helpers/shadow-input'
+import {setup} from '#testHelpers'
+import {getActiveElement, getActiveElementOrBody} from '#src/utils'
+
+describe('focus tests with not shadow root', () => {
```
Comment (ph-fritsche) [inline tests/utils/focus/getActiveElement.ts:5]: Unnecessary describe block. Not all edge cases are necessarily in the binary `with shadow` vs `without shadow` so this might create a legacy of misleading structure.

File: tests/utils/focus/getActiveElement.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+    expect(getActiveElement(document)).toBe(
+      element.shadowRoot?.querySelector('input'),
+    )
+  })
+  test('focus just body', async () => {
+    setup('<shadow-input></shadow-input>', {focus: false})
+
+    expect(getActiveElementOrBody(document)).toBe(document.body)
+  })
```
Comment (ph-fritsche) [inline tests/utils/focus/getActiveElement.ts:31]: This is an anti-pattern. This tests asserts that our test setup is doing nothing while suggesting that has something to do with util being used here.

File: tests/utils/focus/getActiveElement.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+import '../../_helpers/shadow-input'
+import {setup} from '#testHelpers'
+import {getActiveElement, getActiveElementOrBody} from '#src/utils'
+
+describe('focus tests with not shadow root', () => {
+  test('focus input element', async () => {
+    const {element} = setup('<input />')
+
+    expect(getActiveElement(document)).toBe(element)
```
Comment (ph-fritsche) [inline tests/utils/focus/getActiveElement.ts:9]:
```suggestion
    expect(getActiveElementOrBody(document)).toBe(element)
```

We should test through the util that is probably what should be used in all our APIs when retrieving the focused element.

File: tests/utils/focus/getActiveElement.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+  test('focus should be body', async () => {
+    setup('<input />', {focus: false})
+
+    expect(getActiveElementOrBody(document)).toBe(document.body)
+  })
+})
+
+describe('on shadow DOM', () => {
+  test('focus contained input element', async () => {
```
Comment (ph-fritsche) [inline tests/utils/focus/getActiveElement.ts:20]:
```suggestion
  test('get focused element inside shadow tree', async () => {
```

File: src/clipboard/copy.ts [hunk 10]
```diff
@@ -1,10 +1,10 @@
 import {copySelection} from '../document'
-import type {Instance} from '../setup'
-import {writeDataTransferToClipboard} from '../utils'
+import {Instance} from '../setup'
```
[the remark this answers is not in the document]
Comment (Christian24) [reply src/clipboard/copy.ts:2]: This is a mistake. I will fix it.

File: src/clipboard/copy.ts [hunk 11]
```diff
@@ -1,10 +1,10 @@
...
+import {getActiveElementOrBody, writeDataTransferToClipboard} from '../utils'
 
 export async function copy(this: Instance) {
-  const doc = this.config.document
-  const target = doc.activeElement ?? /* istanbul ignore next */ doc.body
+  const doc: Document = this.config.document
```
[the remark this answers is not in the document]
Comment (Christian24) [reply src/clipboard/copy.ts:6]: It is not in my environment. I honestly don't know why but my eslint fails saying that the type is any and therefore it is unsafe.

File: tests/_helpers/shadow-input.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+      this.$input?.setAttribute(name, value)
+    }
+  }
+
+  public get value(): string {
+    return this.$input?.value ?? ''
+  }
+
+  /**
+   * Overwrite focus handling because delegatesFocus is not implemented in jsdom
```
[the remark this answers is not in the document]
Comment (Christian24) [reply tests/_helpers/shadow-input.ts:44]: I don't know. Depends upon what you want to achieve. We used exactly this when Firefox did not yet support `delegatesFocus`. And I know of some libraries which do this too. It is a question of which browser (aka test environment you target). My tests use karma, so overwriting `focus()` is not necessary, `delegatesFocus` would work. However, as a component author if you use `delegatesFocus` or overwrite `focus()` is your choice. I don't think we should make assumptions here, especially ones that wouldn't be true in a real browser. I think this should rather be fixed in jsdom or we should document that `delegatesFocus` is not implemented in jsdom. But I don't think that this is a problem of this library or this library should do something about it. Maybe the comment is misleading?

File: tests/utility/type.ts (hunk 5 above)
Comment (Christian24) [reply tests/utility/type.ts:101]: Maybe this is more related to #1002? The `click` just means it looses focus.

File: tests/utils/focus/getActiveElement.ts (hunk 6 above)
Comment (Christian24) [reply tests/utils/focus/getActiveElement.ts]: Removed it.

File: tests/utils/focus/getActiveElement.ts (hunk 7 above)
Comment (Christian24) [reply tests/utils/focus/getActiveElement.ts]: I tried making it more explicit.

File: tests/utils/focus/getActiveElement.ts (hunk 8 above)
Comment (Christian24) [reply tests/utils/focus/getActiveElement.ts]: I did change all usages to use `getActiveElementOrBody`. Are users expected to use this? I would love to remove the export for `getActiveElement`, but this would be a breaking change, so holding back on that for now.

File: tests/clipboard/copy.ts [hunk 13]
```diff
@@ -1,5 +1,6 @@
 import userEvent from '#src'
 import {render, setup} from '#testHelpers'
+import '../_helpers/shadow-input'
```
[the remark this answers is not in the document]
Comment (Christian24) [reply tests/clipboard/copy.ts]: I made this explicit everywhere.

File: tests/_helpers/setup.ts [hunk 14]
```diff
@@ -28,24 +28,26 @@ export function render<Elements extends Element | Element[] = HTMLElement>(
   if (typeof focus === 'string') {
     ;(assertSingleNodeFromXPath(focus, div) as HTMLElement).focus()
   } else if (focus !== false) {
-    ;(div.querySelector(FOCUSABLE_SELECTOR) as HTMLElement | undefined)?.focus()
+    const element: HTMLElement | null = div.querySelector(FOCUSABLE_SELECTOR)
+    if (element) {
+      element.focus()
+    } else if (div.firstChild) {
+      ;(div.firstChild as HTMLElement).focus()
+    }
```
[the remark this answers is not in the document]
Comment (Christian24) [reply tests/_helpers/setup.ts]: No, I assumed it is a custom element. I changed this to check for firstElementChild and check the tag for `hyphen`, thus making clear it is a custom element and therefore focusable.

File: tests/_helpers/shadow-input.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (Christian24) [reply tests/_helpers/shadow-input.ts]: I changed the comment a bit to make more clear that this is actually functionality of this test component and not a hack.

File: tests/_helpers/setup.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (ph-fritsche) [reply tests/_helpers/setup.ts]: Not every custom element is focusable. And the behavior depends on `ShadowRoot.delegatesFocus` or the overloaded `HTMLElement.focus`.
This util is supposed to provide a good default for our tests that makes it easy for future developers to write tests that focus on the tested feature without polluting the tests with too much arrange logic.
I want to make sure there is nothing subtle about this function. It currently focuses the first focusable element in the rendered fragment by default.
Let's add a function that ensures this behavior when one or more elements with shadow DOM are used here.

File: tests/_helpers/shadow-input.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (ph-fritsche) [reply tests/_helpers/shadow-input.ts]: I just checked in the browser: If I click the `shadow-input` outside the nested `input` (you need to change the style to achieve this, but this doesn't matter here) in the browser, the `document.activeElement` does not change - the `focus` method is not triggered.
Therefore this behavior depends on our test setup calling `focus` on this element which would not be triggered by any user interaction in the browser.

If a workaround for Jsdom not implementing `ShadowRoot.delegatesFocus` is needed, this should be applied when setting up the environment.

File: tests/utility/type.ts (hunk 5 above)
Comment (ph-fritsche) [reply tests/utility/type.ts]: Which is correct. `shadow-input` attaches the shadow DOM with `delegatesFocus: false` . Clicking the `shadow-input` outside of `input` focuses the closest focusable element - which is `document.body` here.
If this is not intended, we need to fix the component and/or our implementation instead of bending the test ;)

#1002 is environment-specific. While it still needs someone to investigate why it happens and how we might be able to fix it, it doesn't affect our tests running in Jsdom.

File: tests/utils/focus/getActiveElement.ts [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+import {setup} from '#testHelpers'
+import {getActiveElementOrBody} from '#src/utils'
+
+test('focused input element', async () => {
+  const {element} = setup('<input />')
+
+  expect(getActiveElementOrBody(document)).toBe(element)
+})
+
+test('focus should be body', async () => {
```
Comment (ph-fritsche) [inline tests/utils/focus/getActiveElement.ts:14]:
```suggestion
test('default to body as active element', async () => {
```

File: tests/utils/focus/getActiveElement.ts [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    )
+  })
+  test('focus just body', async () => {
+    defineShadowInputCustomElementIfNotDefined()
+    const {element} = setup<ShadowInput>('<shadow-input></shadow-input>', {
+      focus: false,
+    })
+    element.blur()
+    expect(getActiveElementOrBody(document)).toBe(document.body)
+  })
```
Comment (ph-fritsche) [inline tests/utils/focus/getActiveElement.ts:36]: It still doesn't test anything about `getActiveElementOrBody`. It merely ensures that after `element.blur()` the test setup is the same as in the other test with `{focus: false}` above.

Comment (Christian24) [reply tests/utils/focus/getActiveElement.ts:36]: What I wanted to test is that nothing is currently focused. I don't see how this is wrong, but I can remove the test if you want.

Comment (ph-fritsche) [reply tests/utils/focus/getActiveElement.ts:36]: The test suggests that `getActiveElementOrBody` should have a specific behavior on this state of the document. But the created state is exactly the initial state which is already covered by a test above. Therefore it just obfuscates the requirements for `getActiveElementOrBody` instead of describing them.

Comment (Christian24) [reply tests/utils/focus/getActiveElement.ts]: Gotcha. I removed the test.

File: src/clipboard/copy.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (Christian24) [reply src/clipboard/copy.ts]: Fixed.

File: tests/_helpers/setup.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Christian24) [reply tests/_helpers/setup.ts]: At work we only have custom elements that are focusable. I just tried this in a variety of browsers and got different results per browser. I am super confused now ;) I will try to write a little helper, so we can figure out the focusable element and call it a day.

[the remark this answers is not in the document]
Comment (ph-fritsche) [reply tests/_helpers/setup.ts]:
```ts
function findFocusable(container: Element | ShadowRoot) {
  for(const el of container.querySelectorAll('*')) {
    if (el.matches(FOCUSABLE_SELECTOR)) {
      return el
    } else if (el.shadowRoot) {
      const f = findFocusable(el.shadowRoot)
      if (f) {
       return f
      }
    }
}
```

File: tests/_helpers/shadow-input.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (Christian24) [reply tests/_helpers/shadow-input.ts]: How would you apply that though?

File: tests/_helpers/setup.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Christian24) [reply tests/_helpers/setup.ts]: I added a helper myself which also skips searching inside the shadowRoot if delegatesFocus is defined.

File: tests/_helpers/shadow-input.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (Christian24) [reply tests/_helpers/shadow-input.ts]: Removed since the helper it is no longer necessary.

File: tests/utility/type.ts (hunk 5 above)
Comment (Christian24) [reply tests/utility/type.ts]: I removed the test. This looked like a quick win, but it is not it seems, so removing this out of this PR.

File: tests/_helpers/setup.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (ph-fritsche) [reply tests/_helpers/setup.ts]: `ShadowRoot.delegatesFocus` can be `false`.
`ShadowRoot.delegatesFocus` can be `true`, but the element could not contain a focusable element.
A focusable element inside a shadow tree might precede or follow on a focusable element outside of it.

File: tests/utility/type.ts [hunk 17]
```diff
@@ -95,11 +95,12 @@ test('do nothing on disabled element', async () => {
...
+    const {element, user} = setup<ShadowInput>('<shadow-input></shadow-input>')
 
-    // Skip click because delegatesFocus is not implemented in jsdom
+    // Skip click because the element is already focused
     await user.type(element, 'test', {skipClick: true})
 
-    expect((element as ShadowInput).value).toEqual('test')
+    expect(element.value).toEqual('test')
   })
 })
```
Comment (ph-fritsche) [inline tests/utility/type.ts:106]: If the test skips the click because it wouldn't work with it, we should add a test describing the missing implementation on `.pointer()` and move the typing test to `.keyboard()`.

File: tests/_helpers/setup.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Christian24) [reply tests/_helpers/setup.ts]: This is true. My idea was to fallback to the engine's implementation if there is one. I think overriding this behavior in general if it exists in the browser engine is a bad idea. It is true though, we will not know what the element will do with the focus.

File: tests/utility/type.ts (hunk 17 above)
Comment (Christian24) [reply tests/utility/type.ts:106]: I removed this, because my primary focus is cut, copy and paste. Maybe this can be part of another PR.

File: tests/_helpers/setup.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Christian24) [reply tests/_helpers/setup.ts]: Done

[the remark this answers is not in the document]
Comment (ph-fritsche) [reply tests/_helpers/setup.ts]: The goal is to provide a state to the document that can be used to test a certain feature.
The util needs to provide the same state in different environments.
We will probably run these tests in Jsdom and in a browser. See #1019

File: tests/utility/type.ts (hunk 17 above)
Comment (ph-fritsche) [reply tests/utility/type.ts]: But the test was there before. We shouldn't just remove it to merge this.
We need to make sure that we have a good understanding of what is missing and needs to be fixed later if we don't deal with it in this PR.

File: tests/_helpers/setup.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Christian24) [reply tests/_helpers/setup.ts]: I am not convinced. This feels like reimplementing the browser. If I want to test pasting support in my custom element for example, I rely on the proper element being focused. What the browser does on `delegatesFocus` is important to get the proper element focused. 

I may want to use karma to run this test in a real browser and I would assume that `delegatesFocus` would be used, because that's why I am running it in a real browser. While in reality what would happen would be that `getFocusableElement` would be used, yet as a user I would be unaware of this. If our implementation differs from the browser's implementation people would never notice this, because we would always use our implementation without them knowing.

File: tests/utility/type.ts (hunk 17 above)
Comment (Christian24) [reply tests/utility/type.ts]: As I said I assumed it was a quick win. It was not, so I removed it for now. It is not part of what this PR describes anyway.

File: tests/_helpers/setup.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (ph-fritsche) [reply tests/_helpers/setup.ts]: > If I want to test pasting support in my custom element for example, I rely on the proper element being focused. What the browser does on delegatesFocus is important to get the proper element focused.

That's exactly the issue here. The setup util provides a certain document state, e.g. `elementA` being focused, so that a test can assert a feature is correctly implemented, e.g. the pasted content being inserted in `elementA`.
If the setup util provides a different document state, e.g. `elementB` being focused, it is expected that pasted content is not being inserted in `elementA`.

Now that `elementA` has to get focused on some interaction is a different feature. On that one it is important that there is a certain document state, e.g. `elementA` not being focused.

If the document state created by the setup util depends on the environment, this might make it very difficult to find the relevant bit when multiple tests break and point at different features.

File: tests/utility/type.ts (hunk 5 above)
Comment (Christian24) [reply tests/utility/type.ts]: I added a new test for `keyboard`.

File: tests/_helpers/setup.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Christian24) [reply tests/_helpers/setup.ts]: Oh, sorry, I was mistaken. `setup` is only really used internally right? Then I agree this is a good idea. Sorry for the confusion. I updated the code.

[the remark this answers is not in the document]
Comment (ph-fritsche) [reply tests/_helpers/setup.ts]: Yes, it's part of the `#testHelpers`. When we implement handling focus on the `mousedown` on the shadow tree, we'll need to provide the same behavior as the browser would.

File: tests/_helpers/setup.ts [hunk 18]
```diff
@@ -105,3 +103,25 @@ export function setup<Elements extends Element | Element[] = HTMLElement>(
...
+      const possibleFocusableChildElement = getFocusableElement(
+        child.shadowRoot,
+      )
+      if (possibleFocusableChildElement) {
+        return possibleFocusableChildElement
+      }
+    }
+  }
+  return null
+}
```
Comment (ph-fritsche) [inline tests/_helpers/setup.ts:127]: Could you explain your preference over:
https://github.com/testing-library/user-event/pull/1033#discussion_r947111604

Comment (Christian24) [reply tests/_helpers/setup.ts:127]: None really. I can also use that if that works better for you. This basically just comes from my tinkering with delegatesFocus.

Comment (ph-fritsche) [reply tests/_helpers/setup.ts:127]: If there is no advantage, i'd prefer the shorter implementation that takes the first focusable regardless of whether it's inside a shadow tree and there are other focusable elements after that.

Comment (Christian24) [reply tests/_helpers/setup.ts]: Done

**Comment by ph-fritsche** [comment]
@Christian24 Thanks for your work on this. I'll review again with fresh eyes and try to figure out if - based on what we learned here - there are some easy improvements towards full shadow DOM support that we could add to that release as well.

**Comment by Christian24** [comment]
> @Christian24 Thanks for your work on this. I'll review again with fresh eyes and try to figure out if - based on what we learned here - there are some easy improvements towards full shadow DOM support that we could add to that release as well.

Thanks for taking the time to review it!

**Comment by Christian24** [comment]
@ph-fritsche is there something I can do to help get this, #1038, #1039, #1040 merged?

**Comment by ph-fritsche** [comment]
@Christian24 I'd like to resolve #1019 before merging the pending PRs so that we have proper testing utils in place to verify our changes and any future bug reports. Also the CI is broken - as it seems due to conflicts with more recent versions of Typescript and Eslint. We need to fix these before we can move forward.
I opened a draft for updating the test environment in #1081 . Any help on this would be highly welcome.

**Comment by Christian24** [comment]
@ph-fritsche Sorry, I got somewhat swallowed. I would like to help. I will try to rebase this to get started.
